models/baseModels/tests/testLoyaltyProgram.spec.ts

```diff
@@ -0,0 +1,275 @@
+
+  t.ok(
+    await fyo.db.exists(
+      ModelNameEnum.LoyaltyPointEntry,
+      loyaltyPointEntryData?.name
+    ),
+    `Loyalty Point Entry '${loyaltyPointEntryData?.name}' should be created against Sales Invoice '${sinvDoc.name}'`
+  );
+
+  const loyaltyProgramDoc = (await fyo.doc.getDoc(
+    ModelNameEnum.LoyaltyProgram,
+    sinvDoc.loyaltyProgram
+  )) as Party;
+
+  const selectedTier: CollectionRulesItems | undefined = getLoyaltyProgramTier(loyaltyProgramDoc, fyo.pesa(itemData.rate));
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyProgramTier,
+    selectedTier?.tierName,
+    `tier must be ${loyaltyPointEntryData?.loyaltyProgramTier} because it is the matched tier'`
+  );
+
+  const tierData: any = collectionRulesData.find((rule) => {
+    return rule.tierName === loyaltyPointEntryData?.loyaltyProgramTier;
+  });
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyPoints,
+    (itemData.rate * tierData?.collectionFactor) as number,
+    `${loyaltyPointEntryData?.loyaltyPoints} loyalty points should be added to customer '${loyaltyPointEntryData?.customer}'`
+  );
+});
+
+test('create Sales Invoice with future date and verify loyalty points are not created', async (t) => {
+  const futureDate = new Date(new Date().setDate(new Date().getDate() + 20));
+
+  const sinvDoc = await createSalesInvoice();
+  sinvDoc.date = futureDate;
+
+  await sinvDoc.sync();
+  await sinvDoc.runFormulas();
+  await sinvDoc.submit();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.name),
+    `Sales Invoice '${sinvDoc.name}' exists`
+  );
+
+  const loyaltyPointEntryData = await loyaltyPointEntryDoc(
+    sinvDoc.name as string
+  );
+
+  t.equals(
+    loyaltyPointEntryData?.name,
+    undefined,
+    'Loyalty points should not be created for a future-dated Sales Invoice'
+  );
+});
+
+test('create Sales Invoice with Loyalty Points redemption and verify negative loyalty point entry is created correctly', async (t) => {
+  const sinvDoc = await createSalesInvoice();
+
+  sinvDoc.redeemLoyaltyPoints = true;
+  sinvDoc.loyaltyPoints = 1000;
+
+  await sinvDoc.sync();
+  await sinvDoc.runFormulas();
+  await sinvDoc.submit();
+
+  t.ok(
+    fyo.db.exists(ModelNameEnum.SalesInvoice, sinvDoc.name),
+    `Sales Invoice '${sinvDoc.name}' exists`
+  );
+
+  t.equals(
+    sinvDoc.loyaltyPoints,
+    1000,
+    `Sales Invoice '${sinvDoc.name}' should have 1000 loyalty points for redemption`
+  );
+
+  const loyaltyPointEntryData = await loyaltyPointEntryDoc(
+    sinvDoc.name as string
+  );
+
+  t.ok(
+    await fyo.db.exists(
+      ModelNameEnum.LoyaltyPointEntry,
+      loyaltyPointEntryData?.name
+    ),
+    `Negative Loyalty Point Entry '${loyaltyPointEntryData?.name}' should be created against Sales Invoice '${sinvDoc.name}'`
+  );
+
+  t.equals(
+    loyaltyPointEntryData?.loyaltyPoints,
+    -1000,
+    `Negative ${loyaltyPointEntryData?.loyaltyPoints} loyalty points should be added to customer '${loyaltyPointEntryData?.customer}'`
```

Comment:
message needs to be changed to something like 
_redeemed loyalty point matches the loyalty points used_